editor/filesystem_dock.cpp

```diff
@@ -1563,10 +1563,37 @@ void FileSystemDock::_try_duplicate_item(const FileOrFolder &p_item, const Strin
 }
 
 void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, String> &p_renames, const HashMap<String, ResourceUID::ID> &p_uids) const {
-	// Update the paths in ResourceUID, so that UIDs remain valid.
-	for (const KeyValue<String, ResourceUID::ID> &pair : p_uids) {
-		if (p_renames.has(pair.key)) {
-			ResourceUID::get_singleton()->set_id(pair.value, p_renames[pair.key]);
+	for (const KeyValue<String, String> &pair : p_renames) {
+		// Update UID path.
+		const HashMap<String, ResourceUID::ID>::ConstIterator I = p_uids.find(pair.key);
+		if (I) {
+			ResourceUID::get_singleton()->set_id(I->value, pair.value);
+		}
+
+		int index = -1;
+		EditorFileSystemDirectory *efd = EditorFileSystem::get_singleton()->find_file(pair.key, &index);
```

Comment:
This is nearly 100% copied from `EditorFileSystem._update_script_classes`. I wonder if we can instead extract the necessary part of the method to a new method and call if from both locations?

The code in `EditorFileSystem` also handles the case when the file does not exist, which looks good to have in place as well.

Reply:
I did consider this, but the differences are enough that trying to extract it into a single function would be more annoying than useful. It's not like its a lot of code.

The "file" should be guaranteed to exist, because it is a `EditorFileSystemDirectory` fetched before a fs rescan is called, it uses data from before the move/rename operation. Can you think of a case where this would break?

Reply:
It makes the maintenance easier, as right now only `EditorFileSystem._update_script_classes` is responsible for updating script classes.
With your PR, there will be 2 places and if something needs to be fixed in the logic, one place may be forgotten.

Is there really a difference? I only see the order of `ScriptServer::remove_global_class_by_path(path)` and `ClassDB::is_parent_class(type, "Script")`, but that is also handled below at the language check. All other differences can be handled easily.

No, but for FS operations generally, it is always better to check and be safe as you can never be 100% sure. 
Your code right now can fail or even crash if the index is not found and therefore -1.